pysph/sph/rb_rotation_matrices.py

```diff
@@ -0,0 +1,542 @@
+        trq = declare('object')
+        body_limits = declare('object')
+        n_body = declare('int')
+        fx = declare('object')
+        fy = declare('object')
+        fz = declare('object')
+        x = declare('object')
+        y = declare('object')
+        z = declare('object')
+        xcm = declare('object')
+        i = declare('int')
+        j = declare('int')
+        i3 = declare('int')
+        l = declare('int')
+        r = declare('int')
+
+        frc = dst.force
+        trq = dst.torque
+        fx = dst.fx
+        fy = dst.fy
+        fz = dst.fz
+        x = dst.x
+        y = dst.y
+        z = dst.z
+        xcm = dst.xcm
+        n_body = dst.n_body[0]
+        body_limits = dst.body_limits
+
+        frc[:] = 0
+        trq[:] = 0
+
+        for i in range(n_body):
+            # left limit of body i
+            l = body_limits[2 * i]
+            # right limit of body i
+            r = body_limits[2 * i + 1]
+
+            # step i
+            i3 = 3 * i
+            for j in range(l, r):
+                frc[i3] += fx[j]
+                frc[i3 + 1] += fy[j]
+                frc[i3 + 2] += fz[j]
+
+                # torque due to force on particle i
+                # (r_i - com) \cross f_i
+                dx = x[j] - xcm[i3]
+                dy = y[j] - xcm[i3 + 1]
+                dz = z[j] - xcm[i3 + 2]
+
+                # torque due to force on particle i
+                # dri \cross fi
+                trq[i3] += (dy * fz[j] - dz * fy[j])
+                trq[i3 + 1] += (dz * fx[j] - dx * fz[j])
+                trq[i3 + 2] += (dx * fy[j] - dy * fx[j])
+
+
+def normalize_orientation(orien):
+    a1 = np.array([orien[0], orien[3], orien[6]])
+    a2 = np.array([orien[1], orien[4], orien[7]])
+    a3 = np.array([orien[2], orien[5], orien[8]])
+    # norm of col0
+    na1 = np.linalg.norm(a1)
+
+    b1 = a1 / na1
+
+    b2 = a2 - np.dot(b1, a2) * b1
+    nb2 = np.linalg.norm(b2)
+    b2 = b2 / nb2
+
+    b3 = a3 - np.dot(b1, a3) * b1 - np.dot(b2, a3) * b2
+    nb3 = np.linalg.norm(b3)
+    b3 = b3 / nb3
+
+    orien[0] = b1[0]
+    orien[3] = b1[1]
+    orien[6] = b1[2]
+    orien[1] = b2[0]
+    orien[4] = b2[1]
+    orien[7] = b2[2]
+    orien[2] = b3[0]
+    orien[5] = b3[1]
+    orien[8] = b3[2]
+
+
+class RK2StepRigidBody(IntegratorStep):
+    def py_initialize(self, dst, t, dt):
+        for i in range(dst.n_body[0]):
+            i3 = 3 * i
+            i9 = 9 * i
+            for j in range(i3, i3 + 3):
+                # save the center of mass and center of mass velocity
+                dst.xcm_0[j] = dst.xcm[j]
+                dst.vcm_0[j] = dst.vcm[j]
+
+                # save the current angular momentum
+                dst.ang_mom_0[j] = dst.ang_mom[j]
+
+            # save the current orientation
+            # print(dst.orientation_R)
+            for j in range(i9, i9 + 9):
+                dst.orientation_R_0[j] = dst.orientation_R[j]
+
+    def initialize(self):
```

Comment:
Instead of doing this, you can simply avoid the need for initialize by writing your own integrator.  This empty initialize is not ideal.